packages/compass-indexes/src/modules/regular-indexes.ts

```diff
@@ -96,61 +86,103 @@ export const INITIAL_STATE: State = {
   error: null,
 };
 
-export default function reducer(state = INITIAL_STATE, action: AnyAction) {
-  if (isAction<IndexesAddedAction>(action, ActionTypes.IndexesAdded)) {
+export default function reducer(
+  state = INITIAL_STATE,
+  action: AnyAction
+): State {
+  if (
+    isAction<FetchIndexesStartedAction>(action, ActionTypes.FetchIndexesStarted)
+  ) {
     return {
       ...state,
-      indexes: action.indexes,
+      error: null,
+      isRefreshing: action.isRefreshing,
     };
   }
 
-  if (isAction<SetIsRefreshingAction>(action, ActionTypes.SetIsRefreshing)) {
+  if (
+    isAction<FetchIndexesSucceededAction>(
+      action,
+      ActionTypes.FetchIndexesSucceeded
+    )
+  ) {
+    // Merge the newly fetched indexes and the existing in-progress ones.
+    const inProgressIndexes = state.inProgressIndexes;
+    const allIndexes = _mergeInProgressIndexes(
+      action.indexes,
+      cloneDeep(inProgressIndexes)
+    );
+
     return {
       ...state,
-      isRefreshing: action.isRefreshing,
+      indexes: allIndexes,
+      isRefreshing: false,
     };
   }
 
-  if (isAction<SetErrorAction>(action, ActionTypes.SetError)) {
+  if (
+    isAction<FetchIndexesFailedAction>(action, ActionTypes.FetchIndexesFailed)
+  ) {
     return {
       ...state,
       error: action.error,
+      indexes: [],
+      isRefreshing: false,
     };
   }
 
   if (
-    isAction<InProgressIndexAddedAction>(
+    isAction<IndexCreationStartedAction>(
       action,
-      ActionTypes.InProgressIndexAdded
+      CreateIndexActionTypes.IndexCreationStarted
     )
   ) {
+    // Add the new in-progress index to the in-progress indexes.
+    const inProgressIndexes = [
+      ...state.inProgressIndexes,
+      action.inProgressIndex,
+    ];
+
+    // Merge the in-progress indexes into the existing indexes.
+    const allIndexes = _mergeInProgressIndexes(
+      state.indexes,
+      cloneDeep(inProgressIndexes)
+    );
+
     return {
       ...state,
-      inProgressIndexes: [...state.inProgressIndexes, action.index],
+      inProgressIndexes,
```

Comment:
Maybe I'm missing something, but isn't this piece of state is just a duplicate of what we'd already have in the `indexes` state anyway and so breaks the single source of truth rule here kinda? Do you think it might be worth considering some alternatives for this state shape so that we don't keep this in two places: either keeping two completely separate and merging only in the view for displaying purposes (the combined list is basically a derived data from two sources), or keeping them in a single list, making it clear in the types that some indexes in the list are "in progress", and doing all the merging over the same list.

Sort of tangentially related (this bug exists on main too and is kinda related to how this merge and the state shape is structured): if I try to create an index with a custom name that matches an already existing index, it will do this merge over an existing index matching by name and will mark it as "failed", which is kinda confusing taking into account that this index actually still exists and is functioning.

Reply:
Yeah I admittedly didn't rethink the in-progress and merging logic. I'm wondering if it should just not be in here at all: Have two lists and merge it later before displaying in the UI. I figured we'd touch that again when we have to merge the rolling indexes in too. ie. have three lists: in-progress rolling indexes, in-progress regular indexes and actual regular indexes, then merge it before we display. ie. don't merge those together in the reducer.

I suppose I can make changes in this PR too.

Reply:
Since we're merging on name I think that real regular indexes should win out over in-progress ones and it should just replace it entirely. Otherwise we'll inevitably have weirdness like what you pointed out.

Reply:
Ah, actually a good point and a nice way to slice this work, we can totally get to it when we are adding rolling indexes to this mix 👍